contracts/origination/OriginationControllerSTIRFRY.sol

```diff
@@ -0,0 +1,268 @@
+     * @param sig                           The loan terms signature, with v, r, s fields, and possible extra data.
+     * @param sigProperties                 Signature nonce and max uses for this nonce.
+     * @param itemPredicates                The predicate rules for the items in the bundle.
+     *
+     * @return loanId                       The unique ID of the new loan.
+     */
+    function initializeStirfryLoan(
+        LoanLibrary.LoanTerms calldata loanTerms,
+        StirfryData calldata stirfryData,
+        address borrower,
+        address lender,
+        Signature calldata sig,
+        SigProperties calldata sigProperties,
+        LoanLibrary.Predicate[] calldata itemPredicates
+    ) external returns (uint256 loanId) {
+        // input validation
+        originationHelpers.validateLoanTerms(loanTerms);
+
+        address vaultAddress = vaultFactory.instanceAt(loanTerms.collateralId);
+        _validateStirfryTerms(loanTerms, stirfryData, vaultAddress);
+
+        {
+            // signature validation
+            Side neededSide = isSelfOrApproved(borrower, msg.sender) ? Side.LEND : Side.BORROW;
+
+            address signingCounterparty = neededSide == Side.LEND ? lender : borrower;
+            address callingCounterparty = neededSide == Side.LEND ? borrower : lender;
+
+            (bytes32 sighash, address externalSigner) = _recoverSignature(loanTerms, sig, sigProperties, neededSide, signingCounterparty, itemPredicates);
+
+            _validateCounterparties(signingCounterparty, callingCounterparty, msg.sender, externalSigner, sig, sighash);
+
+            // consume signer nonce
+            loanCore.consumeNonce(externalSigner, sigProperties.nonce, sigProperties.maxUses);
+        }
+
+        // initialize loan
+        loanId = _initialize(loanTerms, stirfryData, vaultAddress, borrower, lender);
+
+        // Run predicates check at the end of the function, after vault is in escrow. This makes sure
+        // that re-entrancy was not employed to withdraw collateral after the predicates check occurs.
+        if (itemPredicates.length > 0) originationHelpers.runPredicatesCheck(borrower, lender, loanTerms, itemPredicates);
+    }
+
+    // =========================================== HELPERS ==============================================
+
+    /**
+     * @notice Validate that the stirfry data provided is valid for the given loan terms.
+     *
+     * @param loanTerms                     The terms of the loan.
+     * @param stirfryData                   The stirfry data to validate.
+     * @param vaultAddress                  The address of the vault.
+     */
+    function _validateStirfryTerms(
+        LoanLibrary.LoanTerms calldata loanTerms,
+        StirfryData calldata stirfryData,
+        address vaultAddress
+    ) internal view {
+        // verify the vaulted currency is allowed to be paired with the loan terms payable currency
+        bytes32 currencyHash = keccak256(
```

Comment:
Maybe this var can just be called `key` instead of `currencyHash` because it is not really a hash of any currency.